webapp/src/main/java/com/box/l10n/mojito/rest/repository/BranchStatisticWS.java

```diff
@@ -52,41 +44,70 @@ public class BranchStatisticWS {
+        @RequestParam(value = "createdBefore", required = false) DateTime createdBefore,
+        @RequestParam(value = "createdAfter", required = false) DateTime createdAfter,
+        @PageableDefault(sort = "id", direction = Sort.Direction.DESC) Pageable pageable) {
     // Two phase querying: 1. retrieve BranchStatistic IDs for pagination first
-    Page<Long> branchStatisticIds =
-        sparseBranchStatisticRepository.findAllWithIdOnly(
-            where(ifParamNotNull(createdByUserNameEquals(createdByUserName)))
-                .and(ifParamNotNull(branchEquals(branchId)))
-                .and(ifParamNotNull(branchNameEquals(branchName)))
-                .and(ifParamNotNull(search(search)))
-                .and(ifParamNotNull(deletedEquals(deleted)))
-                .and(ifParamNotNull(empty(empty)))
-                .and(ifParamNotNull(createdBefore(createdBefore)))
-                .and(ifParamNotNull(createdAfter(createdAfter)))
-                .and(totalCountLessThanOrEqualsTo(totalCountLte)),
-            pageable);
+    // Retrieve all the Branch Statistic Ids without the totalCountLte filter:
+    Page<Long> branchStatisticIds = sparseBranchStatisticRepository.findAllWithIdOnly(where(
+                    ifParamNotNull(createdByUserNameEquals(createdByUserName))).and(
+                    ifParamNotNull(branchEquals(branchId))).and(
+                    ifParamNotNull(branchNameEquals(branchName))).and(
+                    ifParamNotNull(search(search))).and(
+                    ifParamNotNull(deletedEquals(deleted))).and(
+                    ifParamNotNull(empty(empty))).and(
+                    ifParamNotNull(createdBefore(createdBefore))).and(
+                    ifParamNotNull(createdAfter(createdAfter)))
+            , pageable);
 
-    // 2. Hydrate BranchStatistic entities for JSON response using eager loading to avoid N+1
-    // queries
-    List<BranchStatistic> branchStatistics =
-        branchStatisticRepository.findByIdIn(branchStatisticIds.getContent(), pageable.getSort());
+    // Retrieve the Branch Statistic Ids which have TmTextUnit count less than the 'totalCountLte'
+    Page<Long> branchStatisticIdsWithLessThan = sparseBranchStatisticRepository.findAllWithIdOnly(where(
+                    ifParamNotNull(createdByUserNameEquals(createdByUserName))).and(
+                    ifParamNotNull(branchEquals(branchId))).and(
+                    ifParamNotNull(branchNameEquals(branchName))).and(
+                    ifParamNotNull(search(search))).and(
+                    ifParamNotNull(deletedEquals(deleted))).and(
+                    ifParamNotNull(empty(empty))).and(
+                    ifParamNotNull(createdBefore(createdBefore))).and(
+                    ifParamNotNull(createdAfter(createdAfter))).and(
+                    totalCountLessThanOrEqualsTo(totalCountLte))
+            , pageable);
```

Comment:
Using pageable here will result in results getting out of sync after a couple of pages, once a couple of larger branches are thrown in the mix. Passing in pageable to a query can only be done once for that reasons.

This query should take in branchStatisticIds as a query condition and filter for count based on those results and pageable should not be used here.

Reply:
Yes, I tried that unfortunately the branchStatisticIds doesn't give me the all the list of items as it's paged, instead it would give me only the first page of Ids.

Reply:
How many branches do you expect to return from this API request? More than what one page (the current one) contains?

Reply:
Yes in my test DB I had 16 branches returned, in production if user doesn't select the 'only my branches' it could even return 57 branches (6 pages).

Reply:
When I was testing I was expecting 16 (>30K branches) + 8 (<=30K) branches, so 3 pages was the number I was expecting from the method.

Reply:
If the size of the page passed in is 10 - only 10 results should be returned though - there's no value then to ever query data for more than 10 branches at a time.